In [11]:
pip install librosa

  Using cached librosa-0.8.1-py3-none-any.whl (203 kB)
  Using cached SoundFile-0.10.3.post1-py2.py3.cp26.cp27.cp32.cp33.cp34.cp35.cp36.pp27.pp32.pp33-none-win_amd64.whl (689 kB)
  Using cached pooch-1.5.2-py3-none-any.whl (57 kB)
  Using cached resampy-0.2.2.tar.gz (323 kB)
  Using cached audioread-2.1.9.tar.gz (377 kB)
  Created wheel for audioread: filename=audioread-2.1.9-py3-none-any.whl size=23154 sha256=4cbbdbaadecbac5e271f8ebd21538a9651798bd40a19bfab28dae8edf8bbac08
  Stored in directory: c:\users\gianm\appdata\local\pip\cache\wheels\49\5a\e4\df590783499a992a88de6c0898991d1167453a3196d0d1eeb7
  Created wheel for resampy: filename=resampy-0.2.2-py3-none-any.whl size=320731 sha256=c2de224ad9a43530cd2042efd0809ea787999cac489fd0e46f346408b50ba60d
  Stored in directory: c:\users\gianm\appdata\local\pip\cache\wheels\6f\d1\5d\f13da53b1dcbc2624ff548456c9ffb526c914f53c12c318bb4
Successfully built audioread resampy
Note: you may need to restart the kernel to use updated packages.


In [65]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.api.types import is_string_dtype
from pandas.api.types import is_numeric_dtype
import librosa

# 2.1 Getting your data 

We want to read the datasets to try to understand which key we can use to merge them together.

First of all we want to fill empty values in the datasets.
The below function take as an input a dataframe and check if in the columns there are any missing values of numeric or string types and fills this value with an empty string, in case of strings, or with a mean of the values of that column in case of numbers.

In [66]:
def filler(dataframe):
    for col in dataframe.columns:
        if dataframe[col].isnull().any():
            if is_string_dtype(dataframe[col]):
                dataframe[col] = dataframe[col].fillna("")
            elif is_numeric_dtype(dataframe[col]):
                dataframe[col] = dataframe[col].fillna(dataframe[col].mean())

We take the datasets needed:

In [67]:
tracks = pd.read_csv("tracks.csv")
echonest = pd.read_csv("echonest.csv")
features = pd.read_csv("features.csv")

In [68]:
tracks.head()

track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4        20               0  2008-11-26 01:45:05  2009-01-06 00:00:00   

  album_engineer  album_favorites  album_id  \
0            NaN                4         1   
1            NaN                4         1   
2            NaN                4         1   
3            NaN                4         6   
4            NaN                2         4   

                           album_information  album_listens album_producer  \
0                                    <p></p>           6073            NaN   
1                                    <p></p>           6073            NaN   
2                                    <p></p>           6073            NaN   
3                                        NaN          47632            NaN   
4  <p> "spiritual songs" from Nicky Cook</p>           2710            NaN   

   ... track_information track_interest  track_language_code  \
0  ...               NaN           4656                   en   
1  ...               NaN           1470                   en   
2  ...               NaN           1933                   en   
3  ...               NaN          54881                   en   
4  ...               NaN            978                   en   

                                       track_license track_listens  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1293   
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...           514   
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1151   
3  Attribution-NonCommercial-NoDerivatives (aka M...         50135   
4  Attribution-NonCommercial-NoDerivatives (aka M...           361   

  track_lyricist track_number track_publisher  track_tags      track_title  
0            NaN            3             NaN          []             Food  
1            NaN            4             NaN          []     Electric Ave  
2            NaN            6             NaN          []       This World  
3            NaN            1             NaN          []          Freeway  
4            NaN            3             NaN          []  Spiritual Level  

[5 rows x 53 columns]

In [69]:
tracks.shape

(106574, 53)

In [70]:
echonest.head()

track_id  audio_features_acousticness  audio_features_danceability  \
0         2                     0.416675                     0.675894   
1         3                     0.374408                     0.528643   
2         5                     0.043567                     0.745566   
3        10                     0.951670                     0.658179   
4       134                     0.452217                     0.513238   

   audio_features_energy  audio_features_instrumentalness  \
0               0.634476                         0.010628   
1               0.817461                         0.001851   
2               0.701470                         0.000697   
3               0.924525                         0.965427   
4               0.560410                         0.019443   

   audio_features_liveness  audio_features_speechiness  audio_features_tempo  \
0                 0.177647                    0.159310               165.922   
1                 0.105880                    0.461818               126.957   
2                 0.373143                    0.124595               100.260   
3                 0.115474                    0.032985               111.562   
4                 0.096567                    0.525519               114.290   

   audio_features_valence metadata_album_date  ... temporal_features_214  \
0                0.576661                 NaN  ...             -1.992303   
1                0.269240                 NaN  ...             -1.582331   
2                0.621661                 NaN  ...             -2.288358   
3                0.963590          2008-03-11  ...             -3.662988   
4                0.894072                 NaN  ...             -1.452696   

   temporal_features_215 temporal_features_216  temporal_features_217  \
0               6.805694              0.233070               0.192880   
1               8.889308              0.258464               0.220905   
2              11.527109              0.256821               0.237820   
3              21.508228              0.283352               0.267070   
4               2.356398              0.234686               0.199550   

  temporal_features_218 temporal_features_219  temporal_features_220  \
0              0.027455               0.06408                3.67696   
1              0.081368               0.06413                6.08277   
2              0.060122               0.06014                5.92649   
3              0.125704               0.08082                8.41401   
4              0.149332               0.06440               11.26707   

   temporal_features_221  temporal_features_222  temporal_features_223  
0                3.61288              13.316690             262.929749  
1                6.01864              16.673548             325.581085  
2                5.86635              16.013849             356.755737  
3                8.33319              21.317064             483.403809  
4               11.20267              26.454180             751.147705  

[5 rows x 250 columns]

In [71]:
echonest.shape

(13129, 250)

In [72]:
features.head()

track_id  chroma_cens_kurtosis_01  chroma_cens_kurtosis_02  \
0         2                 7.180653                 5.230309   
1         3                 1.888963                 0.760539   
2         5                 0.527563                -0.077654   
3        10                 3.702245                -0.291193   
4        20                -0.193837                -0.198527   

   chroma_cens_kurtosis_03  chroma_cens_kurtosis_04  chroma_cens_kurtosis_05  \
0                 0.249321                 1.347620                 1.482478   
1                 0.345297                 2.295201                 1.654031   
2                -0.279610                 0.685883                 1.937570   
3                 2.196742                -0.234449                 1.367364   
4                 0.201546                 0.258556                 0.775204   

   chroma_cens_kurtosis_06  chroma_cens_kurtosis_07  chroma_cens_kurtosis_08  \
0                 0.531371                 1.481593                 2.691455   
1                 0.067592                 1.366848                 1.054094   
2                 0.880839                -0.923192                -0.927232   
3                 0.998411                 1.770694                 1.604566   
4                 0.084794                -0.289294                -0.816410   

   chroma_cens_kurtosis_09  ...  tonnetz_std_04  tonnetz_std_05  \
0                 0.866868  ...        0.054125        0.012226   
1                 0.108103  ...        0.063831        0.014212   
2                 0.666617  ...        0.040730        0.012691   
3                 0.521217  ...        0.074358        0.017952   
4                 0.043851  ...        0.095003        0.022492   

   tonnetz_std_06  zcr_kurtosis_01  zcr_max_01  zcr_mean_01  zcr_median_01  \
0        0.012111         5.758890    0.459473     0.085629       0.071289   
1        0.017740         2.824694    0.466309     0.084578       0.063965   
2        0.014759         6.808415    0.375000     0.053114       0.041504   
3        0.013921        21.434212    0.452148     0.077515       0.071777   
4        0.021355        16.669037    0.469727     0.047225       0.040039   

   zcr_min_01  zcr_skew_01  zcr_std_01  
0    0.000000     2.089872    0.061448  
1    0.000000     1.716724    0.069330  
2    0.000000     2.193303    0.044861  
3    0.000000     3.542325    0.040800  
4    0.000977     3.189831    0.030993  

[5 rows x 519 columns]

In [73]:
features.shape

(106574, 519)

We can notice that a common column is **track_id**, so let's check if there are any missing values:

In [74]:
tracks["track_id"].isnull().any()

False

In [75]:
echonest["track_id"].isnull().any()

False

In [76]:
features["track_id"].isnull().any()

False

Given that there are no missing values in "**track_id**" columns in any dataset, we choose to use it as the key for the merge of the datasets.

I build a copy and work on it to don't damage the original files.

In [77]:
tracks_c = tracks.copy()
echonest_c = echonest.copy()
features_c = features.copy()

In [78]:
# First we are gonna fill the datasets using the function filler()
filler(tracks_c)
filler(echonest_c)
filler(features_c)

In [79]:
tracks_c.head()

track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4        20               0  2008-11-26 01:45:05  2009-01-06 00:00:00   

  album_engineer  album_favorites  album_id  \
0                               4         1   
1                               4         1   
2                               4         1   
3                               4         6   
4                               2         4   

                           album_information  album_listens album_producer  \
0                                    <p></p>           6073                  
1                                    <p></p>           6073                  
2                                    <p></p>           6073                  
3                                                     47632                  
4  <p> "spiritual songs" from Nicky Cook</p>           2710                  

   ... track_information track_interest  track_language_code  \
0  ...                             4656                   en   
1  ...                             1470                   en   
2  ...                             1933                   en   
3  ...                            54881                   en   
4  ...                              978                   en   

                                       track_license track_listens  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1293   
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...           514   
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1151   
3  Attribution-NonCommercial-NoDerivatives (aka M...         50135   
4  Attribution-NonCommercial-NoDerivatives (aka M...           361   

  track_lyricist track_number track_publisher  track_tags      track_title  
0                           3                          []             Food  
1                           4                          []     Electric Ave  
2                           6                          []       This World  
3                           1                          []          Freeway  
4                           3                          []  Spiritual Level  

[5 rows x 53 columns]

In [80]:
echonest_c.head()

track_id  audio_features_acousticness  audio_features_danceability  \
0         2                     0.416675                     0.675894   
1         3                     0.374408                     0.528643   
2         5                     0.043567                     0.745566   
3        10                     0.951670                     0.658179   
4       134                     0.452217                     0.513238   

   audio_features_energy  audio_features_instrumentalness  \
0               0.634476                         0.010628   
1               0.817461                         0.001851   
2               0.701470                         0.000697   
3               0.924525                         0.965427   
4               0.560410                         0.019443   

   audio_features_liveness  audio_features_speechiness  audio_features_tempo  \
0                 0.177647                    0.159310               165.922   
1                 0.105880                    0.461818               126.957   
2                 0.373143                    0.124595               100.260   
3                 0.115474                    0.032985               111.562   
4                 0.096567                    0.525519               114.290   

   audio_features_valence metadata_album_date  ... temporal_features_214  \
0                0.576661                      ...             -1.992303   
1                0.269240                      ...             -1.582331   
2                0.621661                      ...             -2.288358   
3                0.963590          2008-03-11  ...             -3.662988   
4                0.894072                      ...             -1.452696   

   temporal_features_215 temporal_features_216  temporal_features_217  \
0               6.805694              0.233070               0.192880   
1               8.889308              0.258464               0.220905   
2              11.527109              0.256821               0.237820   
3              21.508228              0.283352               0.267070   
4               2.356398              0.234686               0.199550   

  temporal_features_218 temporal_features_219  temporal_features_220  \
0              0.027455               0.06408                3.67696   
1              0.081368               0.06413                6.08277   
2              0.060122               0.06014                5.92649   
3              0.125704               0.08082                8.41401   
4              0.149332               0.06440               11.26707   

   temporal_features_221  temporal_features_222  temporal_features_223  
0                3.61288              13.316690             262.929749  
1                6.01864              16.673548             325.581085  
2                5.86635              16.013849             356.755737  
3                8.33319              21.317064             483.403809  
4               11.20267              26.454180             751.147705  

[5 rows x 250 columns]

In [81]:
features_c.head()

track_id  chroma_cens_kurtosis_01  chroma_cens_kurtosis_02  \
0         2                 7.180653                 5.230309   
1         3                 1.888963                 0.760539   
2         5                 0.527563                -0.077654   
3        10                 3.702245                -0.291193   
4        20                -0.193837                -0.198527   

   chroma_cens_kurtosis_03  chroma_cens_kurtosis_04  chroma_cens_kurtosis_05  \
0                 0.249321                 1.347620                 1.482478   
1                 0.345297                 2.295201                 1.654031   
2                -0.279610                 0.685883                 1.937570   
3                 2.196742                -0.234449                 1.367364   
4                 0.201546                 0.258556                 0.775204   

   chroma_cens_kurtosis_06  chroma_cens_kurtosis_07  chroma_cens_kurtosis_08  \
0                 0.531371                 1.481593                 2.691455   
1                 0.067592                 1.366848                 1.054094   
2                 0.880839                -0.923192                -0.927232   
3                 0.998411                 1.770694                 1.604566   
4                 0.084794                -0.289294                -0.816410   

   chroma_cens_kurtosis_09  ...  tonnetz_std_04  tonnetz_std_05  \
0                 0.866868  ...        0.054125        0.012226   
1                 0.108103  ...        0.063831        0.014212   
2                 0.666617  ...        0.040730        0.012691   
3                 0.521217  ...        0.074358        0.017952   
4                 0.043851  ...        0.095003        0.022492   

   tonnetz_std_06  zcr_kurtosis_01  zcr_max_01  zcr_mean_01  zcr_median_01  \
0        0.012111         5.758890    0.459473     0.085629       0.071289   
1        0.017740         2.824694    0.466309     0.084578       0.063965   
2        0.014759         6.808415    0.375000     0.053114       0.041504   
3        0.013921        21.434212    0.452148     0.077515       0.071777   
4        0.021355        16.669037    0.469727     0.047225       0.040039   

   zcr_min_01  zcr_skew_01  zcr_std_01  
0    0.000000     2.089872    0.061448  
1    0.000000     1.716724    0.069330  
2    0.000000     2.193303    0.044861  
3    0.000000     3.542325    0.040800  
4    0.000977     3.189831    0.030993  

[5 rows x 519 columns]

Once we got all the datasets well filled, we can start merging them using as key the **track_id** column.

In [82]:
final = tracks_c.merge(features_c.merge(echonest_c, on = "track_id"), on = "track_id")
print(final.shape)
final.head()

(13129, 820)


track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4       134               0  2008-11-26 01:44:45  2009-01-05 00:00:00   

  album_engineer  album_favorites  album_id album_information  album_listens  \
0                               4         1           <p></p>           6073   
1                               4         1           <p></p>           6073   
2                               4         1           <p></p>           6073   
3                               4         6                            47632   
4                               4         1           <p></p>           6073   

  album_producer  ... temporal_features_214 temporal_features_215  \
0                 ...             -1.992303              6.805694   
1                 ...             -1.582331              8.889308   
2                 ...             -2.288358             11.527109   
3                 ...             -3.662988             21.508228   
4                 ...             -1.452696              2.356398   

   temporal_features_216 temporal_features_217 temporal_features_218  \
0               0.233070              0.192880              0.027455   
1               0.258464              0.220905              0.081368   
2               0.256821              0.237820              0.060122   
3               0.283352              0.267070              0.125704   
4               0.234686              0.199550              0.149332   

  temporal_features_219 temporal_features_220 temporal_features_221  \
0               0.06408               3.67696               3.61288   
1               0.06413               6.08277               6.01864   
2               0.06014               5.92649               5.86635   
3               0.08082               8.41401               8.33319   
4               0.06440              11.26707              11.20267   

   temporal_features_222 temporal_features_223  
0              13.316690            262.929749  
1              16.673548            325.581085  
2              16.013849            356.755737  
3              21.317064            483.403809  
4              26.454180            751.147705  

[5 rows x 820 columns]

As we can see, the number of rows is almost 13k, as demanded.

## 2.2 Choose your features (variables)!

We notice that the number of features is very big (820 columns!) and we want to reduce the number of variables we have to work with.

So let's apply dimensionality reduction!